### PR TITLE
Feature calendar booking

### PR DESCRIPTION
**Pull request by Emcrowz**
Basic functionality of the booking feature using calendar. NOT using a modal.
Addition of the '**react-calendar**' [[link](https://www.npmjs.com/package/react-calendar)] library to the project.

Other changes/implementations:
- User needs to be logged to make a booking
- Calendar and time selection from business defined work times
- API integration for sending Booking record to DB

### CODE REVIEW

File: backend/src/Models/Booking.ts [hunk 1]
```diff
@@ -1,50 +1,56 @@
...
+  _id: Types.ObjectId;
   businessId: Types.ObjectId;
-  date: Date;
-  time: string;
+  userId: Types.ObjectId;
   userEmail: string;
-  userName: string;
-  status: "confirmed" | "pending" | "cancelled";
+  reservationTime: Array<number>;
+  // time: string;
```
Comment (rokasandreikenas) [inline backend/src/Models/Booking.ts:9]: ?

File: backend/src/Models/Booking.ts [hunk 2]
```diff
@@ -1,50 +1,56 @@
...
+const bookingSchema = new mongoose.Schema<IBooking>(
+  {
+    businessId: {
+      type: mongoose.Schema.Types.ObjectId,
+      required: true,
+    },
+    // time: {
+    //   type: String,
+    //   required: [true, "field is required. e.g. 14:00"], // Time must be provided
+    // },
```
Comment (rokasandreikenas) [inline backend/src/Models/Booking.ts:22]: ?

File: frontend/src/Components/Booking/Booking.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export interface Booking {
+  businessId: string;
+  userId: string;
+  userEmail: string;
+  reservationTime: Array<number>;
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/Booking.ts:5]: naudojam `number[]`

File: frontend/src/Components/Booking/BookingPannel.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
+import { useState } from "react";
+import Calendar from "react-calendar";
+// import "react-calendar/dist/Calendar.css"; // built-in calendar style
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/BookingPannel.tsx:3]: ?

File: frontend/src/Components/Booking/BookingPannel.tsx [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import { Button } from "../Common/Button";
+import { fetchBookings, postBooking } from "./BookingApi";
+import { useQuery } from "@tanstack/react-query";
+import "./CalendarStyles.css";
+
+interface BookingPannelProps {
+  bookingBusiness: string;
+  bookingUser: User;
+  officialWorkTime: Array<number>;
+  workTimes: Array<Array<number>>;
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/BookingPannel.tsx:15]: [] naudojam

File: frontend/src/Components/Booking/BookingPannel.tsx [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+type ValuePiece = Date | null;
+type Value = ValuePiece | [ValuePiece, ValuePiece];
+
+const useBookings = () => {
+  return useQuery({
+    queryKey: ["BOOKINGS"],
+    queryFn: fetchBookings,
+  });
+};
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/BookingPannel.tsx:27]: iskeliam

File: frontend/src/Components/Booking/BookingPannel.tsx [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      <div style={{ margin: "0.5rem 0" }}>
+        {currentBookings.map(
+          (booking) =>
+            bookingBusiness === booking.businessId && (
+              <p>
+                {booking?.userEmail} | {booking?.reservationDate.toLocaleString()}
+              </p>
+            ),
+        )}
+      </div> */}
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/BookingPannel.tsx:91]: ?

File: frontend/src/Components/Booking/BookingPannel.tsx [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+          (workTime) =>
+            checkForAvailableBooking(workTime) && (
+              <Button onClick={() => handleTimeSelection(workTime[0], workTime[1])}>
+                {workTime[0]}:{workTime[1] === 0 ? "00" : workTime[1]}
+              </Button>
+            ),
+        )}
+      </div>
+
+      <div style={{ display: "flex", justifyContent: "space-around", textAlign: "center" }}>
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/BookingPannel.tsx:110]: no inline styling

File: frontend/src/Components/Booking/BookingPannel.tsx [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        onClickDay={(datePicked) => setTimePicked(datePicked)}
+        value={datePicked}
+      />
+
+      <div className={styles.workTimeButtons}>
+        {workTimes?.map(
+          (workTime) =>
+            checkForAvailableBooking(workTime) && (
+              <Button onClick={() => handleTimeSelection(workTime[0], workTime[1])}>
+                {workTime[0]}:{workTime[1] === 0 ? "00" : workTime[1]}
```
Comment (rokasandreikenas) [inline frontend/src/Components/Booking/BookingPannel.tsx:104]: gal iskelti i utilts, nes labai sunkiai skaitosi kas cia daroma

File: frontend/src/Components/Business/Business.ts [hunk 10]
```diff
@@ -7,4 +7,6 @@ export interface Business {
   contactPerson: string;
   email: string;
   imageUrls: string[];
+  officialWorkingTime: Array<number>;
+  workTimes: Array<Array<number>>;
```
Comment (rokasandreikenas) [inline frontend/src/Components/Business/Business.ts:11]: []

File: frontend/src/Pages/BusinessDetails.tsx [hunk 11]
```diff
@@ -1,17 +1,21 @@
...
 import axios from "axios";
 import styles from "./BusinessDetails.module.css";
 import { Button } from "../Components/Common/Button";
 import { BusinessList } from "../Components/Business/BusinessList";
+import { BookingPannel } from "../Components/Booking/BookingPannel";
+import { UserContext } from "../Components/Context/UserContext";
 
 export const BusinessDetails = () => {
   const [business, setBusiness] = useState<Business | null>();
+  const [isBooking, setIsBooking] = useState(false);
```
Comment (rokasandreikenas) [inline frontend/src/Pages/BusinessDetails.tsx:16]: isBookingOpen

**Comment by Emcrowz** [comment]
Design and other changes pushed this PR into the back. Changes made from here are transfered to the #20 PR. 
This branch is closed.

**Pull request closed without merging**